Introduce "Feature States" for managing snapshots of system indices

AthenaEryma opened the pull request:
This PR expands the meaning of `include_global_state` for snapshots to include system indices. If `include_global_state` is `true` on creation, system indices will be included in the snapshot regardless of the contents of the `indices` field. If `include_global_state` is `true` on restoration, system indices will be restored (if included in the snapshot), regardless of the contents of the `indices` field. Index renaming is not applied to system indices, as system indices rely on their names matching certain patterns. If restored system indices are already present, they are automatically deleted prior to restoration from the snapshot to avoid conflicts.

This behavior can be overridden to an extent by including a new field in the snapshot creation or restoration call, `feature_states`, which contains an array of strings indicating the "feature" for which system indices should be snapshotted or restored. For example, this call will only restore the `watcher` and `security` system indices (in addition to `index_1`):

```
POST /_snapshot/my_repository/snapshot_2/_restore
{
  "indices": "index_1",
  "include_global_state": true,
  "feature_states": ["watcher", "security"]
}
```

If `feature_states` is present, the system indices associated with those features will be snapshotted or restored regardless of the value of `include_global_state`. All system indices can be omitted by providing a special value of `none` (`"feature_states": ["none"]`), or included by omitting the field or explicitly providing an empty array (`"feature_states": []`), similar to the `indices` field.

The list of currently available features can be retrieved via a new "Get Snapshottable Features" API:
```
GET /_snapshottable_features
```

which returns a response of the form:
```
{
    "features": [
        {
            "name": "tasks",
            "description": "Manages task results"
        },
        {
            "name": "kibana",
            "description": "Manages Kibana configuration and reports"
        }
    ]
}
```

Features currently map one-to-one with `SystemIndexPlugin`s, but this should be considered an implementation detail. The Get Snapshottable Features API and snapshot creation rely upon all relevant plugins being installed on the master node.

Further, the list of feature states included in a given snapshot is exposed by the Get Snapshot API, which now includes a new field, `feature_states`, which contains a list of the feature states and their associated system indices which are included in the snapshot. All system indices in feature states are also included in the `indices` array for backwards compatibility, although explicitly requesting system indices included in a feature state is deprecated. For example, an excerpt from the Get Snapshot API showing `feature_states`:
```
"feature_states": [
    {
        "feature_name": "tasks",
        "indices": [
            ".tasks"
        ]
    }
],
"indices": [
    ".tasks",
    "test1",
    "test2"
]
```

Relates https://github.com/elastic/elasticsearch/issues/61657

AthenaEryma committed:
Add plugin name to SystemIndexPlugin

AthenaEryma committed:
Use plugin name for system index descriptor map

AthenaEryma committed:
Add pluginStates to create and restore snapshot requests

AthenaEryma committed:
Plumb system index descriptor map into SnapshotsService

AthenaEryma committed:
Remove unnecessary comment

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Only include pluginStates in request XContent if non-empty

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Add plugin states to SnapshotInfo

AthenaEryma committed:
pluginStates->featureStates

AthenaEryma committed:
More plugin->feature

AthenaEryma committed:
Break out SnapshotFeatureInfo, add tests and equals/hashcode

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Null guard for `featureStates` in `toXContent`

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Another null guard for `featureStates` in `toXContent`

AthenaEryma committed:
writeTo version guard

AthenaEryma committed:
ACTUAL writeTo version guard

AthenaEryma committed:
SnapshotInfo featureStates now non-nullable

AthenaEryma committed:
Test featureStates in CreateSnapshotResponseTests & fix parser

AthenaEryma committed:
NOCOMMIT -> GWB todos

AthenaEryma committed:
Line length

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Add `feature_states` to expected responses in the docs

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Add system indices during snapshot creation (first draft)

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Address various TODOs

AthenaEryma committed:
Refactor giant pile of streams

AthenaEryma committed:
Spotless

AthenaEryma committed:
Line length

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Fix accidental change to startTime and add featureStates to Entry serialization tests

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Fix NPE in SnapshotResiliencyTests

AthenaEryma committed:
Remove null guards for SnapshotInfo featureStates as featureStates cannot be null

AthenaEryma committed:
Omit empty featureStates when creating the snapshots

AthenaEryma committed:
Ensure built-in features (i.e. tasks) are propagated to SnapshotsService

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Don't include global state in test snapshot (for now) to prevent conflicts

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Very basic first cut of snapshottable features API

williamrandolph committed:
Add tests for snapshot and restore of system indices

AthenaEryma committed:
Add feature description to system index plugins

AthenaEryma committed:
Fix typo in test name

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

williamrandolph committed:
Merge branch 'si/snapshot-and-restore' into si/snapshot-tests

williamrandolph committed:
Delete indices by UUID rather than name

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Plumbing through descriptions

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

williamrandolph committed:
Allow restoring by feature state

williamrandolph committed:
Code review fixes

AthenaEryma committed:
Merge pull request #1 from williamrandolph/si/snapshot-tests

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Treat a `null` and empty list of feature states differently

A `null` means "default to `include_global_state`" - i.e. restore all
system indices if it's `true`, or none otherwise. An empty list means
restore *no* system indices, even if `include_global_state` is `true`.

AthenaEryma committed:
Don't rename system indices

AthenaEryma committed:
Refactoring/cleanup. No functionality changes

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Name and description for Fleet

williamrandolph committed:
Add deprecation warning for directly restoring system indices

williamrandolph committed:
Add test for bad feature state request

williamrandolph committed:
Make HashSet variable final

AthenaEryma committed:
Deprecate direct system index restoration

This commit covers two goals:

* Emit a deprecation log when a request explicitly restores a system index (rather than via feature state) if the snapshot has feature states
* Throw an exception if a feature state is requested but it's not in the snapshot

The `getFeatureStatesToRestore` method already throws an exception if a requested feature state isn't in the snapshot, so I added an integration test to cover this case.

If one of the indices in the restore request is a system index, it might be from a feature that wasn't snapshotted, so we won't know that it's a system index until we've resolved the index against repository data. We can check that each system index has not been directly requested as we build the list of system indices to delete.

AthenaEryma committed:
Only warn when explicitly requesting system indices if the snapshot is from a version with feature states

AthenaEryma committed:
Some missed s/plugin/feature/

AthenaEryma committed:
Address a couple minor TODO comments

AthenaEryma committed:
Address TODO by generating random feature infos in test

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
One missed s/plugin/feature/

AthenaEryma committed:
Basic associated indices implementation

AthenaEryma committed:
Remove unused method

AthenaEryma committed:
Don't include associated indices in the feature state (but snapshot them)

We don't control the namespace for hidden indices as much as we do for
system indices, so we shouldn't just stomp on any existing indices that
conflict with associated indices. But we can at least ensure they're
included in the snapshot.

AthenaEryma committed:
Delete all system indices for features that are being restored

AthenaEryma committed:
Always restore aliases for system indices

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Don't delete all indices if there aren't any feature states to restore

AthenaEryma committed:
Null guards for FeatureStates

AthenaEryma committed:
Make Get Snapshottable Features non-operator only

AthenaEryma committed:
Fix test now that system indices aren't always deleted

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Name/description for test plugins

williamrandolph committed:
Improve system index snapshot IT

I did some refactoring and cleanup of the test class, and removed some
unnecessary operations from the tests so it's easier to see what they're
doing. I also added comments.

We have four new tests here as well: 

1.`testSystemIndicesCannotBeRenamed` - verify that even if a system
index matches a rename pattern, it isn't renamed in a restore operation
2. `testRestoreSystemIndicesAsGlobalStateWithNullFeatureStateList` -
with `restoreGlobalState` set to true, explicitly pass a `null` for the
list of feature states and verify that all feature states are restored.
3. `testRestoreSystemIndicesAsGlobalStateWithEmptyListOfFeatureStates` -
again restoring global state, pass an empty list for the list of feature
states and verify that no feature states are restored.
4. `testRestoreSystemIndicesAsGlobalStateWithEmptyListOfFeatureStatesNoIndicesSpecified`
- test for BWC edge case behavior. If no indices are specified and an
empty list of feature states is passed, verify that we try to restore
the system index because it was included in the expansion of "all
indices." This behavior will change in the future.

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Add a test to ensure system indices in a feature state are removed when restoring that state

AthenaEryma committed:
Add test to ensure system index aliases are always restored

AthenaEryma committed:
Missed a necessary refresh

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

williamrandolph committed:
Fix rest test and add test coverage elsewhere

williamrandolph committed:
Ignore global state in searchable snapshot test

AthenaEryma committed:
Nest features array under an object key in Get Features API to better match API standards

AthenaEryma committed:
Get Snapshottable Features API spec & yaml spec/smoke test

AthenaEryma committed:
Update docs for feature states

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge remote-tracking branch 'gwbrown/si/snapshot-and-restore' into si/snapshot-and-restore

AthenaEryma committed:
Add version-based skip to Get Features YAML test

AthenaEryma committed:
Add HLRC support for Get Features API

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Checkstyle

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

jaymode commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java:
```diff
@@ -389,4 +391,43 @@ public Cancellable deleteAsync(DeleteSnapshotRequest deleteSnapshotRequest, Requ
+
+    /**
+     * Get a list of features which can be included in a snapshot as feature states.
+     * See [link here]
```
Let's either fill this in or remove it and open an issue to update it

jaymode commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java:
```diff
@@ -389,4 +391,43 @@ public Cancellable deleteAsync(DeleteSnapshotRequest deleteSnapshotRequest, Requ
+    }
+
+    /**
+     *
```
Add a blurb about what this does?

jaymode commented on docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:
```diff
@@ -102,10 +102,25 @@ The cluster state includes:
 * Legacy index templates
 * Ingest pipelines
 * {ilm-init} lifecycle policies
+* Data stored in system indices, such as Watches and task records (configurable via `feature_states`)
```
There isn't a great place to leave this comment but this is in the section about what the `cluster state` contains. IMO we should change the wording above to say global state instead of cluster state. Alternatively we can move this out of being included in the cluster state and say that global state also includes data stored in system indices.

jaymode commented on docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:
```diff
@@ -102,10 +102,25 @@ The cluster state includes:
+available for inclusion in the snapshot and their descriptions be can be
+retrieved using the <<get-snapshottable-features-api,get snapshottable features API>>.
+Each feature state includes one or more system indices containing data necessary
+for the function of that feature. Providing an empty array will include no feature
```
I have concerns regarding the consistency of what an empty array means. For indices within a CreateSnapshotRequest, it means all open indices. The behavior for features states is different and this feels wrong to me.

gwbrown replied:
That's fair and a good point - my intent with making it that way was to be able to say "I want to restore my cluster state, and no system indices". Do you have any suggestions for an alternate way of doing that?

Alternatively, is that even something we need at all? This feels like something that should be possible given you can omit features, but I don't necessarily have a use case in mind.

jaymode replied:
A few items that I can think of would be introducing a special `none` option, completely omitting `feature_states`, or `+` `-` and `*` support.

Regarding whether we need it, I guess there could be a case where someone just wants to restore how they did previously and only include items from the cluster state.

jaymode replied:
I agree that the `none` option is the best and also agree that it must be the only item in the array in that case.

jaymode commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -115,6 +115,15 @@ List of <<data-streams,data streams>> included in the snapshot.
+List of feature states which were included when the snapshot was taken,
+including the list of system indices included as part of the feature state. The
+`feature_name` field of each can be used in the `feature_states` parameter when
+restoring the snapshot to restore a subset of feature states. Only present if
```
Is the part about only being present if one or more feature states are included? In the example below, an empty array is shown.

jaymode commented on docs/reference/snapshot-restore/apis/restore-snapshot-api.asciidoc:
```diff
@@ -134,10 +134,22 @@ The cluster state includes:
 * Legacy index templates
 * Ingest pipelines
 * {ilm-init} lifecycle policies
+* For snapshots taken after 7.12.0, data stored in system indices, such as Watches and task records, replacing any existing configuration (configurable via `feature_states`)
```
same comment regarding the docs indicating that this exists in the cluster state.

jaymode commented on docs/reference/snapshot-restore/apis/restore-snapshot-api.asciidoc:
```diff
@@ -134,10 +134,22 @@ The cluster state includes:
+A comma-separated list of feature states you wish to restore. Each feature state contains one or more system indices. The list of feature states
+available in a given snapshot are returned by the <<get-snapshot-api-feature-states, Get Snapshot API>>. Note that feature
+states restored this way will completely replace any existing configuration, rather than returning an error if the system index already exists.
+Providing an empty array will restore no feature states, regardless of the value of `include_global_state`.
```
Is empty array behavior consistent with other aspects of this API?

jaymode commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -32,6 +32,9 @@ By default, all data streams and indices in the snapshot are restored, but the c
 `include_global_state` to `true` in the restore request body.
 
+Because all indices in the snapshot are restored by default, all system indices will be restored
+by default as well,
```
```suggestion
by default as well.
```

jaymode commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -88,7 +91,7 @@ POST /_snapshot/my_backup/snapshot_1/_restore
 
 <1> By default, `include_global_state` is `false`, meaning the snapshot's
-cluster state is not restored.
+cluster state and any feature states are not restored.
```
```suggestion
cluster state and feature states are not restored.
```

jaymode reviewed (commented):
I did not make it far into reviewing, but I think some of the feedback is valuable enough so that we keep moving on this work.

jaymode commented on server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -614,6 +658,88 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        List<String> requestedDataStreams = filterIndices(snapshotInfo.dataStreams(), requestIndices.toArray(String[]::new),
+            IndicesOptions.fromOptions(true, true, true, true));
+        if (requestedDataStreams.isEmpty()) {
+            dataStreams = new HashMap<>();
```
`Map.of()` or `Collections.emptyMap()`?

jaymode commented on server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -614,6 +658,88 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            .filter(featureName -> systemIndices.getFeatures().containsKey(featureName))
+            .collect(Collectors.toList());
+        if (featuresNotOnThisNode.isEmpty() == false) {
+            logger.warn("while restoring feature states from snapshot [{}], some requested feature states are not present on this node: {}",
```
shouldn't this be an error and not a log message?

gwbrown replied:
Looking at this again, I agree, it probably should be an error. Not certain why I didn't make it one originally. I'll change this, although I'm not sure if it's possible to write a test for this case given how our test infrastructure works.

jaymode commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:
```diff
@@ -1170,6 +1170,16 @@ static long machineMemoryFromStats(OsStats stats) {
 
+    @Override
+    public String getFeatureName() {
+        return "machine_learning";
```
I wonder if this should be `ml`? @droberts195 any thoughts?

jaymode reviewed (commented):
I finally made it through this PR. Aside from the comments and as we discussed elsewhere, we should have some tests for how this interacts with security.

AthenaEryma committed:
Cluster state -> global state in docs where it makes sense

AthenaEryma committed:
Fix Client javadoc

AthenaEryma committed:
Typo/wording fixes per review

AthenaEryma committed:
Remove excess line

AthenaEryma committed:
new HashMap<>() -> Collections.emptyMap()

AthenaEryma committed:
Throw an error when attempting to restore feature states that go with features which are not installed

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Clean up error message

AthenaEryma committed:
Flip incorrect boolean logic (d'oh)

gwbrown commented:
ci/2 failure appears to be https://github.com/elastic/elasticsearch/issues/67620

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Add category for deprecation warning

AthenaEryma committed:
Handle null & empty feature states the same in Create/Restore requests

AthenaEryma committed:
Add "none" special case for including no feature states

Also removes `[]` as a way to specify no feature states

AthenaEryma committed:
Don't include feature_states in restore request JSON if it's empty

If we include it, some BWC tests use new_version request classes to make
requests against old_version clusters.

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Exercise feature_states in HLRC tests

williamrandolph committed:
Add integration test for feature states and security index

AthenaEryma committed:
Add a test to ensure system indices can be renamed if restored by name

williamrandolph committed:
Add UUID comparison to security index state change listeners

gwbrown commented:
Opened https://github.com/elastic/elasticsearch/issues/67945 for the BWC failure, I think the other one is due to the changes here (specifically, a snapshot is picking more shards now than expected due to including system indices in snapshots by default). Fix coming shortly.

AthenaEryma committed:
Exclude all feature states in `createSnapshot`

Otherwise some tests can fail because the snapshot includes a different
number of shards than the test expects, due to system indices being
implicitly included.

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge remote-tracking branch 'gwbrown/si/snapshot-and-restore' into si/snapshot-and-restore

jaymode commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -19,6 +19,43 @@
 
 package org.elasticsearch.indices;
 
+import static java.util.Collections.emptyList;
```
did something change with formatting the imports in your IDE? I've always seen static imports as going last

gwbrown replied:
Fascinating. I didn't intentionally change anything, I'll look into that and see if I can undo whatever happened there.

jaymode reviewed (approved):
LGTM. I left one comment about some import statement re-ordering that is mixed in here.

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

# Conflicts:
#	server/src/main/java/org/elasticsearch/snapshots/RestoreService.java
#	server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexUpgradeServiceTests.java

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -257,6 +268,44 @@ public ClusterState execute(ClusterState currentState) {
                 // Store newSnapshot here to be processed in clusterStateProcessed
                 List<String> indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request));
 
+                List<SnapshotFeatureInfo> featureStates = Collections.emptyList();
```
I think we need BwC here. We shouldn't be triggering ta snapshot that includes `featureStates` as long as there's old master nodes in the cluster that don't understand `featureStates` yet. Even if it might work out magically now (because we redundantly store what to snapshot in `indices` it's a potential point of failure where we move to a different state machine on master failover.

gwbrown replied:
To clarify, what you're recommending here is not using `featureStates` in a mixed cluster where some master nodes are <`FEATURE_STATES_VERSION`?

In that case, f there's a request that explicitly uses feature states, we can just throw an exception that that feature isn't allowed in mixed clusters. But if we have a create snapshot request that implicitly uses feature states (`include_global_state: true`), should we use the pre-7.12.0 behavior regardless of the current version of the master? I don't like silently using the previous behavior like that, but it seems like we'd still want to be able to take a snapshot in that case.

original-brownbear replied:
>  the pre-7.12.0 behavior regardless of the current version of the master?

I think that's the correct approach as long as there's old version nodes in the cluster. At least, it's what we currently do for all the snapshot features/changes, we always keep using the old state machine as long as there's nodes of an old version around. It's just the by far easiest BwC behavior to reason about.

> In that case, f there's a request that explicitly uses feature states, we can just throw an exception that that feature isn't allowed in mixed clusters

Yea I'd throw here if the request explicitly requests feature states. If it doesn't specify anything, just using the previous behavior seems the way to go.

original-brownbear reviewed (commented):
Sorry again for the delay here Gordon. A couple of points. I didn't get through the restore service in full today but there were a couple of fixes to it over the last two days. Might be a good idea to merge `master` in here and make sure all new tests still pass after the changes as well.

AthenaEryma committed:
Undo import changes (thanks IntelliJ)

AthenaEryma committed:
! -> == false

AthenaEryma committed:
DRY up test per review

AthenaEryma committed:
Don't pre-size arrays per review

AthenaEryma committed:
Rearrange `startedEntry` params per review

AthenaEryma committed:
Move `FEATURE_STATES_VERSION` to `SnapshotsService` per review

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -115,8 +118,8 @@ public static Entry startedEntry(Snapshot snapshot, boolean includeGlobalState,
         return new SnapshotsInProgress.Entry(snapshot, true, false, State.STARTED, indices, Collections.emptyList(),
-                startTime, repositoryStateId, ImmutableOpenMap.of(), null, Collections.emptyMap(), version, source,
-                ImmutableOpenMap.of());
+            Collections.emptyList(), startTime, repositoryStateId, ImmutableOpenMap.of(), null, Collections.emptyMap(), version, source,
```
This seems like a good point. Snapshot cloning is relatively new and I don't know much about it - if that's the case though, why are data streams `Collections.emptyList()` here? Shouldn't they be copied as well?

original-brownbear replied:
:facepalm: yet another datastream curve-ball :) This should be fixed as well. Let's do what you did here then maybe and add a TODO for this. I don't think it's critical to fix in this PR since it's broken already for datastreams. Thanks for noticing!

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
"Partial" fix in SnapshotsService#finalizeSnapshotEntry

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
New license headers that didn't get auto-updated

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Don't put Descriptor patterns through IndexNameExpressionResolver

SystemIndexDescriptor patterns can use full Lucene regex syntax, so we
can't feed them to IndexNameExpressionResolver anymore. This commit
introduces a method to SystemIndexDescriptor to resolve all indices
matching a descriptor.

AthenaEryma committed:
Unused imports

AthenaEryma committed:
Only use feature_states if all nodes are on a version that can handle it (BWC)

AthenaEryma committed:
If any feature state indices are partial, drop associated feature state

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Unused import

gwbrown commented:
@original-brownbear I believe I've addressed all of your review comments, with the exception of https://github.com/elastic/elasticsearch/pull/63513#discussion_r570380604, where I interpreted your last comment as saying "let's handle this in a follow-up" - I would have handled it here, but we don't appear to have access to the source snapshot `SnapshotInfo` at the time we create the clone entry.

For convenience, the nontrivial commits addressing your comments are:
- BWC Handling: [`d503868` (#63513)](https://github.com/elastic/elasticsearch/pull/63513/commits/d503868f626644fba39769fc01f6351e5ca6db11)
- Partial-snapshot handling, in which if any shards in a feature state are missing we drop the feature state to prevent it being restored: [`f2d7bbe` (#63513)](https://github.com/elastic/elasticsearch/pull/63513/commits/f2d7bbe390db9e79434a3a165acf6a60f157e83d)

The rest of the new commits are simple style fixes or cleanup, with the exception of [`12d1b25` (#63513)](https://github.com/elastic/elasticsearch/pull/63513/commits/12d1b25775cefb66d682e4d9b62af2826ba635d3) that's more related to system indices infrastructure than snapshots, and I've already run past Jay.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1218,6 +1306,29 @@ private void finalizeSnapshotEntry(SnapshotsInProgress.Entry entry, Metadata met
+        });
+
+        // Now remove any feature states which contain any of those indices, as the feature state is not intact and not safely restorable
+        return entry.featureStates().stream()
```
I think you need to do this the other way around. If you have a partial snapshot, the indices just disappear from the `Entry` instead of showing up as failure. So if indices are missing, this will just quietly ignore those. You can just run the logic the other way around and check that all the indices you need for a feature are in the indices list.

gwbrown replied:
I'm not entirely sure what you mean here - I tried replacing this method with:
```java
final Set<String> indicesInSnapshot = entry.indices().stream()
    .map(IndexId::getName)
    .collect(Collectors.toUnmodifiableSet());

return entry.featureStates().stream()
    .filter(stateInfo -> indicesInSnapshot.containsAll(stateInfo.getIndices()))
    .collect(Collectors.toList());
```

And the test I added to check partial snapshots failed - the index which was only partially captured was still in `entry.indices()`. What I did was apply both approaches (see [this commit](https://github.com/elastic/elasticsearch/pull/63513/commits/38022ddf0fb242b40f239c7398e887f2ee05edf0)) - check for any indices which had failed shards and remove any feature states which include those indices AND check for any indices in feature states which aren't in `entry.indices()`, but I couldn't figure out how to force that case in a test.

Is that what you mean? Or did I misunderstand you?

original-brownbear replied:
@gwbrown 

> Is that what you mean?

Yea exactly we gotta check both. Sorry I missed that you did that correctly in that commit :) Looks good.
Can we still add a test for this please?

> check for any indices in feature states which aren't in entry.indices(), but I couldn't figure out how to force that case in a test.

This should be easy:

1. Start `partial=true` snapshot and block it somewhere
2. Wait for block
3. Delete index
4. Let snapshot continue
5. The finalisation should go through fine and and the feature should not become part of the snapshot.

That should work?

gwbrown replied:
@original-brownbear 

I've written a test that follows that pattern, and it passes, but I'm concerned because if I comment out this line, which applies the logic you recommended of checking `entry.indices()`:
https://github.com/elastic/elasticsearch/blob/90bf18dfa85e4c2a8c896e914d56062ee42ea56c/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1354

...then the test *still passes*, which it sounds like it shouldn't. As far as I can tell, when this test runs, the `featureStates` field in the initial `Entry` doesn't get populated until after the index is deleted, so that feature state is never included in entry to begin with.

I've tried blocking on data nodes instead of blocking the master (see the commented-out lines [removed in this commit](https://github.com/elastic/elasticsearch/pull/63513/commits/d3005866c05e13a7ec7df20824e1680b39a737b9)), and that version also passes, and again still passes if I comment out the same line - in that case, at least one shard fails and the feature state is removed via the logic I originally had in place.

I also tried using, instead of `blockNodeOnAnyFiles`:
- `blockMasterOnWriteIndexFile`, which results in the snapshot completing with all shards successful
- `blockMasterFromFinalizingSnapshotOnSnapFile`, which results in the snapshot entirely failing (`SnapshotMissingException` when attempting to get the `SnapshotInfo`)
- `blockMasterFromFinalizingSnapshotOnIndexFile`, which has the same results as `blockMasterFromFinalizingSnapshotOnSnapFile`

So I'm still confused about how to actually get the original situation you describe, where we can get a feature state in the `Entry` that references an index not in `entry.indices()`.

If you have bandwidth, I'd be happy to talk synchronously on this.

gwbrown replied:
I talked with @original-brownbear synchronously and figured out what was going on. It turned out that with some recent snapshot optimizations, it's more difficult to trigger the case he originally described.

In order to trigger the case reliably, the test now:
- After creating the system indices, starts a new data node
- Creates an index allocated only on the new data node
- Blocks the new data node and starts a snapshot
- Waits for the block to be hit
- Deletes the index
- Unblocks the data node

This allows all shards of the system index to be snapshotted, so that there are no failed shards, but still have that index removed from the snapshot.

There was also something else I misunderstood - when the index is deleted, it's removed from the final `SnapshotInfo`, but not the `Entry` - the final list of indices is generated via `ShardGenerations`. So the method to filter feature states now takes the final list of indices as well as the `Entry` and checks against that, rather than against `Entry#indices()` as the latter does not change.

original-brownbear reviewed (commented):
Looks good @gibrown , just one open point I think. Thanks for the iterations!

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Check indices in snapshot as well as failed shards

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Remove extraneous line break

AthenaEryma committed:
`!` -> `== false`

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Add test - but this test doesn't fail if logic is removed

This test does not get a state where the SnapshotsInProgress.Entry has a feature state but not the index in that feature state

AthenaEryma committed:
Remove commented lines

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore

AthenaEryma committed:
Fix test and use real final indices list

gwbrown commented:
I talked with @original-brownbear elsewhere and confirmed he's okay with merging this once the test is working properly, which it now is. So I'll be merging this once we get a green CI run.

AthenaEryma committed:
Merge branch 'master' into si/snapshot-and-restore
